build: try to enable io_uring if it is not specified

tchaikov opened the pull request:
* configure.py: do not specify an option if "value_when_none" is None
* configure.py: make IO_URING an nullable option, so we don't specify
  this option in the cmake command line option if user does not
  explicity specify it in the configure.py command line.
* cmake: add a function `check_nullable_option()` to extract the
  handling of nullable option out, so it can be reused in future.
* cmake: add the `Seastar_IO_URING` option only if it is specified,
  and try to enable it if it is not specified.

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

nyh commented on CMakeLists.txt:
```diff
@@ -188,6 +188,27 @@ set (Seastar_ALLOC_PAGE_SIZE
   STRING
   "Override the Seastar allocator page size, in bytes.")
 
+function (check_nullable_option option_name package_name)
```
I think the name "check_nullable_option" doesn't really describe what this function is doing. What it seems to really be doing is to check whether a package exists - and allows us to fail if is explictly enabled (--enable-...) or pretend the package doesn't exist even if it does when explicitly disabled (--disable-dpdk).
Would be nice to also get a comment explaining what this function does, in addition to a better name.

tchaikov replied:
thank you for your review.

> I think the name "check_nullable_option" doesn't really describe what this function is doing. 

i knew. i tried but stopped at the 4th name . the first 3 of them were "maybe_set_option", "enable_option_if_possible", "check_option_with_package". would appreciate if if any of us could help suggest a better name.

> What it seems to really be doing is to check whether a package exists - and allows us to fail if is explictly enabled (--enable-...) or pretend the package doesn't exist even if it does when explicitly disabled (--disable-dpdk).
> Would be nice to also get a comment explaining what this function does, in addition to a better name.

quite close. but it does not pretend the package doesn't exist -- it does not set the `<PackageName>_FOUND`. that'd be more confusing than setting the `WITH_<FEATURE>` variable on behalf of user.

sure, will add a comment.

tchaikov replied:
how about `try_to_enable_option_if_not_defined()`?

nyh replied:
I probably completely misunderstood the code, so don't take any advice from me on the naming :-) I thought I was seeing code that checks whether or not the package is found - and then verifies this fact's interaction with the option (e.g., if the package exists but the user asked not to use it, it won't).

tchaikov replied:
yeah. that's exactly what it did in the version you commented and it still has the same behavior in the merged version.

tchaikov commented:
@nyh hi Nadav, instead of trying to be smart, i just added a smaller helper to do the job. while looking at `SeastarDependencies.cmake`, i think there is another opportunity to improve the building system, so i just went ahead and implemented it.

 could you help take another look?

nyh commented on CMakeLists.txt:
```diff
@@ -188,6 +188,17 @@ set (Seastar_ALLOC_PAGE_SIZE
+    return ()
+  endif ()
+  if (${package_name}_FOUND)
+    set (${option_name} ON CACHE BOOL "")
```
Sorry for saying this (and of course it's not relevant to your patch ;-)), but this cmake language is really ugly. This is so weird. One of the common complaints about autoconf was its weird m4 language. So cmake went ahead and invented the only language which is even weirder? :-)

tchaikov replied:
are you referencing `${package_name}_FOUND` ? or the `set()` command? i wanted to use the `package_name` as the parameter for better readability from the caller's perspective, as i think we will repeat the pattern at least for the DPDK support. regarding the latter, yeah, guess, i am used to it already =)

tchaikov commented:
thank you for your review @nyh , if we believe that we should apply the same pattern to more "better to have" option. i'd be happy to create another pull request to do so. some candidates are 

- hwloc
- dpdk

tchaikov commented:
> > thank you for your review @nyh , if we believe that we should apply the same pattern to more "better to have" options. i'd be happy to create another pull request to do so. some candidates are
> > ```
> > * hwloc
> > ```
> 
> If I understand correctly, all modern distros will have hwloc, so perhaps hwloc should no longer be optional at all? Even if it's still optional, if hardly anyone will ever lack hwloc, it's not very important what happens when its missing. So nobody (?) complained about the current situation.
> 

yeah, since nobody is complaining. and as you pointed out, hwloc should always be available on modern distros, so we (Seastar developers) should be safe.

> > ```
> > * dpdk
> > ```
> 
> With dpdk, I think the situation is different. It cannot be "not available" because it's always available as a submodule (right?). Sure, someone can forget to "git submodule update --init", but they shouldn't really... For DPDK I think the question is different - whether it should be enabled by default at all, regardless of whether it's available or not.

ahh, i must have misinterpreted Alec's mail over the discussion group. after revisiting his mail, probably it's safe to keep as it is -- namely, disabled by default.

thanks for your insights!
